src/System.Net.Security/src/System/Net/Unix/SecuritySafeHandles.cs

```diff
@@ -0,0 +1,181 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.Win32.SafeHandles;
+
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Security.Authentication;
+using System.Security.Authentication.ExtendedProtection;
+using System.Security.Cryptography.X509Certificates;
+
+namespace System.Net.Security
+{
+#if DEBUG
+    internal sealed class SafeFreeCertContext : DebugSafeHandle
+    {
+#else
+    internal sealed class SafeFreeCertContext : SafeHandle
+    {
+#endif
+        public SafeFreeCertContext(IntPtr handle) : base(handle, true)
+        {
+        }
+
+        public override bool IsInvalid
+        {
+            get
+            {
+                return IntPtr.Zero == handle;
+            }
+        }
+
+        protected override bool ReleaseHandle()
+        {
+            if (IntPtr.Zero != handle)
+            {
+                Interop.libcrypto.X509_free(handle);
+            }
+            return true;
+        }
+    }
+
+    //
+    // Implementation of handles dependable on FreeCredentialsHandle
+    //
+#if DEBUG
+    internal sealed class SafeFreeCredentials : DebugSafeHandle
+    {
+#else
+    internal sealed class SafeFreeCredentials : SafeHandle
+    {
+#endif
+        private SafeX509Handle _certHandle;
+        private SslProtocols _protocols = SslProtocols.None;
+        private EncryptionPolicy _policy;
+
+        internal SafeX509Handle CertHandle
+        {
+            get { return _certHandle; }
+        }
+
+        internal SslProtocols Protocols
+        {
+            get { return _protocols; }
+        }
+
+        public SafeFreeCredentials(X509Certificate certificate, SslProtocols protocols, EncryptionPolicy policy)
+            : base(IntPtr.Zero, true)
+        {
+            if (null != certificate)
+            {
+                _certHandle = Interop.libcrypto.X509_dup(certificate.Handle);
+            }          
+            _protocols = protocols;
+            _policy = policy;
+        }
+
+        public override bool IsInvalid
+        {
+            get { return SslProtocols.None == _protocols; }
```

Comment:
Did you mean that constructor is called with SslProtocols.None? That won't happen but probably an Assert is in order.

I didn't understand the Dispose requirement. We don't want to dispose the cert (and private key) as long as references are alive. In fact the caller (GenerateToken) calls Dispose on this handle assuming refcount has been bumped up.